src/Session.cc

```diff
@@ -426,6 +428,71 @@ static void remap_shared_mmap(AutoRemoteSyscalls& remote, EmuFs& emu_fs,
   remote.infallible_syscall(syscall_number_for_close(remote.arch()), remote_fd);
 }
 
+const AddressSpace::Mapping& Session::create_shared_mmap(
+    AutoRemoteSyscalls& remote, size_t size, remote_ptr<void> map_hint) {
+  static int nonce = 0;
+  // Create the segment we'll share with the tracee.
+  char path[PATH_MAX];
+  snprintf(path, sizeof(path) - 1, SYSCALLBUF_SHMEM_PATH_PREFIX "%d-%d",
+           remote.task()->real_tgid(), nonce++);
+
+  // Let the child create the shmem block and then send the fd back to us.
+  // This lets us avoid having to make the file world-writeable so that
+  // the child can read it when it's in a different user namespace (which
+  // would be a security hole, letting other users abuse rr users).
+  int child_shmem_fd;
+  {
+    AutoRestoreMem child_path(remote, path);
+    // skip leading '/' since we want the path to be relative to the root fd
+    child_shmem_fd = remote.infallible_syscall(
+        syscall_number_for_openat(remote.task()->vm()->arch()),
+        RR_RESERVED_ROOT_DIR_FD, child_path.get() + 1,
+        O_CREAT | O_EXCL | O_RDWR | O_CLOEXEC, 0600);
+  }
+
+  /* Remove the fs name so that we don't have to worry about
+   * cleaning up this segment in error conditions. */
+  unlink(path);
+
+  ScopedFd shmem_fd = remote.retrieve_fd(child_shmem_fd);
+  resize_shmem_segment(shmem_fd, size);
+  LOG(debug) << "created shmem segment " << path;
+
+  // Map the segment in ours and the tracee's address spaces.
+  void* map_addr;
+  int prot = PROT_READ | PROT_WRITE;
+  int flags = MAP_SHARED;
+  if ((void*)-1 == (map_addr = mmap(nullptr, size, prot, flags, shmem_fd, 0))) {
+    FATAL() << "Failed to mmap shmem region";
+  }
+  if (!map_hint.is_null()) {
+    flags |= MAP_FIXED;
+  }
+  remote_ptr<void> child_map_addr = remote.infallible_mmap_syscall(
+      map_hint, size, prot, flags, child_shmem_fd, 0);
+
+  struct stat st;
+  ASSERT(remote.task(), 0 == ::fstat(shmem_fd, &st));
+  remote.task()->vm()->map(child_map_addr, size, prot, flags, 0, path,
+                           st.st_dev, st.st_ino, nullptr, nullptr, map_addr);
+
+  shmem_fd.close();
+  remote.infallible_syscall(syscall_number_for_close(remote.arch()),
+                            child_shmem_fd);
+  return remote.task()->vm()->mapping_of(child_map_addr);
+}
+
+// Recreate an mmap region that is shared between rr and the tracee
+void Session::recreate_shared_mmap(AutoRemoteSyscalls& remote,
+                                   const AddressSpace::Mapping& m) {
+  assert(m.local_addr != nullptr);
```

Comment:
Yes, that's a different function (in the next commit). This function is the one that recreates the shared mapping upon a clone. It is simpler since we don't have to remap the region to be able to copy it's data.